### PR TITLE
21760 stage 1 overdue ARs notification

### PR DESCRIPTION
**Pull request by kzdev420**
*Issue #:* /bcgov/entity#21760

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: queue_services/entity-emailer/src/entity_emailer/email_processors/ar_overdue_stage_1_notification.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    substitute_template_parts,
+)
+
+
+def process(email_info: dict, token: str) -> dict:  # pylint: disable=too-many-locals, , too-many-branches
+    """Build the email for Involuntary dissolution notification."""
+    logger.debug('ar_overdue_stage_1_notification: %s', email_info)
+    # get business
+    business_identifier = email_info['identifier']
+    business = Business.find_by_identifier(business_identifier)
```
Comment (argush3) [inline queue_services/entity-emailer/src/entity_emailer/email_processors/ar_overdue_stage_1_notification.py:37]: the business identifier is included but the furnishing id in the incoming queue message is how I intended data to be retrieved.  All other info included the queue message is just for easy debugging purposes.  

i.e. retrieve the furnishing id via the furnshings model which should get you to the business model. 

either way, you need the furnishing record to get the email address anyways.

File: queue_services/entity-emailer/src/entity_emailer/email_processors/ar_overdue_stage_1_notification.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    # get extra provincials array
+    extra_provincials = get_extra_provincials(jurisdictions_response)
+    html_out = jnja_template.render(
+        business=business.json(),
+        entity_dashboard_url=get_entity_dashboard_url(business_identifier, token),
+        extra_provincials=extra_provincials
+    )
+    # get recipients
+    recipients = []
+    recipients.append(get_recipient_from_auth(business_identifier, token))  # business email
```
Comment (argush3) [inline queue_services/entity-emailer/src/entity_emailer/email_processors/ar_overdue_stage_1_notification.py:55]: as per my message above, get the email recipient from the `furnishings.email` field.

Comment (kzdev420) [reply queue_services/entity-emailer/src/entity_emailer/email_processors/ar_overdue_stage_1_notification.py]: Updated

File: queue_services/entity-emailer/src/entity_emailer/email_processors/ar_overdue_stage_1_notification.py (hunk 1 above)
Comment (kzdev420) [reply queue_services/entity-emailer/src/entity_emailer/email_processors/ar_overdue_stage_1_notification.py]: Updated

File: queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py [hunk 3]
```diff
@@ -61,6 +62,14 @@ def get_message_context_properties(queue_msg: nats.aio.client.Msg):
...
             message_id = f'{etype}_{filing_id}'
             return create_message_context_properties(etype, message_id, None, identifier, False)
+        if etype == 'bc.registry.dissolution':
+            identifier = email_msg.get('identifier', None)
+            furnishing_id = email_msg['data']['furnishing']['furnishingId']
+            business = ar_overdue_stage_1_notification.get_business_by_furnishing_id(furnishing_id)
+            if business.legal_type in ['BC', 'ULC', 'CC', 'BEN']:
+                source = email_msg.get('source', None)
+                return create_message_context_properties(etype, message_id, source, identifier, False)
+            return create_message_context_properties(etype, None, None, None, False)
```
Comment (argush3) [inline queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py:72]: This is not required as we are not sending the queue message to the emailer from the filer.  

When it's sent from the filer, we are using the already existing format which does not provide a unique "id" that can be used.  This why everything except `etype` == "bc.registry.names.request" is constructing a unique id based off of the incoming properties of a queue message.

We can just extend the check above for `if etype == 'bc.registry.names.request':` to include a check for `bc.registry.dissolution`

File: queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py [hunk 4]
```diff
@@ -171,16 +180,30 @@ def start_tracking_message(message_context_properties: dict, email_msg: dict, ex
...
+def complete_tracking_message(tracker_msg: MessageProcessing, email_msg):
     """Update existing message state to COMPLETED."""
     update_message_status_to_complete(tracker_msg)
 
+    # Update corresponding furnishings entry as PROCESSED
+    etype = email_msg.get('type', None)
+    if etype and etype == 'bc.registry.dissolution':
+        furnishing_id = email_msg['data']['furnishing']['furnishingId']
+        ar_overdue_stage_1_notification.update_furnishing_status(furnishing_id,
+                                                                 Furnishing.FurnishingStatus.PROCESSED)
```
Comment (argush3) [inline queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py:192]: this should not be part of tracker code.

it should take place in the processor

File: queue_services/entity-emailer/src/entity_emailer/email_processors/ar_overdue_stage_1_notification.py
Comment (argush3) [inline queue_services/entity-emailer/src/entity_emailer/email_processors/ar_overdue_stage_1_notification.py:1]: please rename this processor to `involuntary_dissolution_stage_1_notification.py`

File: queue_services/entity-emailer/src/entity_emailer/email_templates/AR_OVERDUE_STAGE_1.html
Comment (argush3) [inline queue_services/entity-emailer/src/entity_emailer/email_templates/AR_OVERDUE_STAGE_1.html:1]: Please rename file  to `INVOL-DIS-STAGE-1.html`

File: queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py [hunk 5]
```diff
@@ -171,16 +180,30 @@ def start_tracking_message(message_context_properties: dict, email_msg: dict, ex
...
                                     existing_tracker_msg: MessageProcessing,
                                     error_details: str):
     """Create a new message with FAILED status or update an existing message to FAILED status."""
+    # Update corresponding furnishings entry as FAILED
+    etype = email_msg.get('type', None)
+    if etype and etype == 'bc.registry.dissolution':
+        furnishing_id = email_msg['data']['furnishing']['furnishingId']
+        ar_overdue_stage_1_notification.update_furnishing_status(furnishing_id,
+                                                                 Furnishing.FurnishingStatus.FAILED)
+
```
Comment (argush3) [inline queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py:206]: this should not be part of tracker code.  

should take place in processor

File: queue_services/entity-emailer/src/entity_emailer/worker.py [hunk 6]
```diff
@@ -142,6 +143,9 @@ def process_email(email_msg: dict, flask_app: Flask):  # pylint: disable=too-man
         elif etype and etype == 'bc.registry.bnmove':
             email = bn_notification.process_bn_move(email_msg, token)
             send_email(email, token)
+        elif etype and etype == 'bc.registry.dissolution':
```
Comment (argush3) [inline queue_services/entity-emailer/src/entity_emailer/worker.py:146]: can we add a check to confirm the `data.furnishingName` is one of the following:

- DISSOLUTION_COMMENCEMENT_NO_AR
- DISSOLUTION_COMMENCEMENT_NO_TR
- DISSOLUTION_COMMENCEMENT_NO_AR_XPRO
- DISSOLUTION_COMMENCEMENT_NO_TR_XPRO

File: queue_services/entity-emailer/tests/unit/test_worker.py [hunk 7]
```diff
@@ -466,3 +469,42 @@ def test_send_email_with_incomplete_payload(app, session, email_msg):
...
+    furnishing = create_furnishing(business)
+    
+    mocker.patch(
+        'entity_emailer.email_processors.ar_overdue_stage_1_notification.get_jurisdictions',
+        return_value=[])
+
+    # run worker
+    with patch.object(AccountService, 'get_bearer_token', return_value=token):
+        with patch.object(worker, 'send_email', return_value='success') as mock_send_email:
+            with patch.object(ar_overdue_stage_1_notification, 'get_recipient_from_auth', return_value='test@test.com'):
```
Comment (argush3) [inline queue_services/entity-emailer/tests/unit/test_worker.py:488]: don't think this is required anymore if we get the email recipient from the furnishings record

File: queue_services/entity-emailer/tests/unit/test_worker.py [hunk 8]
```diff
@@ -466,3 +469,42 @@ def test_send_email_with_incomplete_payload(app, session, email_msg):
...
+                    'source': 'furnishingsJob',
+                    'id': 'f36e3af7-90c3-4859-a6f6-2feefbdc1e37',
+                    'time': '',
+                    'datacontenttype': 'application/json',
+                    'identifier': business_identifier,
+                    'data': {
+                        'furnishing': {
+                            'type': 'PROCESSING',
+                            'furnishingId': furnishing.id,
+                            'furnishingName': furnishing.furnishing_name
```
Comment (argush3) [inline queue_services/entity-emailer/tests/unit/test_worker.py:501]: please verify, the processor only processes the four furnishing types i mentioned in a comment above

File: queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py (hunk 5 above)
Comment (kzdev420) [reply queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py:206]: I am not sure how to detect the HTTP status of notify API in the Processor. 😢
So, I implemented this feature in update_message_tracker

Comment (argush3) [reply queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py:206]: I think what you can do is add a try/catch block around `send_email` in `worker.py`

image

If there is no exception thrown, you can assume it sent the email successfully. 
 if an error is thrown then you can mark it as failed and re-throw the error so the original error handling logic can run.

In both instances, maybe a new function named something like `involuntary_dissolution_stage_1_processor.post_process` can be called

File: queue_services/entity-emailer/src/entity_emailer/email_processors/ar_overdue_stage_1_notification.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+)
+
+
+def process(email_info: dict, token: str) -> dict:  # pylint: disable=too-many-locals, , too-many-branches
+    """Build the email for Involuntary dissolution notification."""
+    logger.debug('ar_overdue_stage_1_notification: %s', email_info)
+    # get business
+    furnishing_id = email_info['data']['furnishing']['furnishingId']
+    furnishing = Furnishing.find_by_id(furnishing_id)
+    business = get_business_by_furnishing_id(furnishing_id)
```
Comment (argush3) [inline queue_services/entity-emailer/src/entity_emailer/email_processors/ar_overdue_stage_1_notification.py:38]: please create a separate PR and add a relationship to the furnishing model so you can just access the business using `furnishing.business`.

File: queue_services/entity-emailer/src/entity_emailer/email_processors/ar_overdue_stage_1_notification.py
Comment (kzdev420) [reply queue_services/entity-emailer/src/entity_emailer/email_processors/ar_overdue_stage_1_notification.py:1]: Updated

File: queue_services/entity-emailer/src/entity_emailer/email_templates/AR_OVERDUE_STAGE_1.html
Comment (kzdev420) [reply queue_services/entity-emailer/src/entity_emailer/email_templates/AR_OVERDUE_STAGE_1.html:1]: Updated

File: queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py (hunk 5 above)
Comment (kzdev420) [reply queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py:206]: Updated

File: queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py (hunk 4 above)
Comment (kzdev420) [reply queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py:192]: Updated

File: queue_services/entity-emailer/src/entity_emailer/worker.py (hunk 6 above)
Comment (kzdev420) [reply queue_services/entity-emailer/src/entity_emailer/worker.py:146]: Updated

File: queue_services/entity-emailer/tests/unit/test_worker.py (hunk 7 above)
Comment (kzdev420) [reply queue_services/entity-emailer/tests/unit/test_worker.py:488]: updated

File: queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py (hunk 3 above)
Comment (kzdev420) [reply queue_services/entity-emailer/src/entity_emailer/message_tracker/tracker.py:72]: Updated

File: queue_services/entity-emailer/tests/unit/test_worker.py (hunk 8 above)
Comment (kzdev420) [reply queue_services/entity-emailer/tests/unit/test_worker.py:501]: Updated

**Pull request closed without merging**